zerver/lib/scim.py

```diff
@@ -0,0 +1,314 @@
+from typing import Any, Callable, Dict, List, Optional, Type, Union
+
+import django_scim.constants as scim_constants
+import django_scim.exceptions as scim_exceptions
+from django.conf import settings
+from django.core.validators import ValidationError, validate_email
+from django.db import models, transaction
+from django.http import HttpRequest
+from django_scim.adapters import SCIMUser
+from scim2_filter_parser.attr_paths import AttrPath
+
+from zerver.lib.actions import (
+    do_change_user_delivery_email,
+    do_create_user,
+    do_deactivate_user,
+    do_reactivate_user,
+)
+from zerver.lib.email_validation import email_allowed_for_realm
+from zerver.lib.request import RequestNotes
+from zerver.lib.subdomains import get_subdomain
+from zerver.models import (
+    DisposableEmailError,
+    DomainNotAllowedForRealmError,
+    EmailContainsPlusError,
+    UserProfile,
+    get_user_by_delivery_email,
+)
+
+
+class ZulipSCIMUser(SCIMUser):
+    id_field = "id"
+
+    def __init__(self, obj: UserProfile, request: Optional[HttpRequest] = None) -> None:
+        # We keep the function signature from the superclass, but this actually
+        # shouldn't be called with request being None.
+        assert request is not None
+
+        # self.obj is populated appropriately by django-scim2 views with
+        # an instance of UserProfile - either fetched from the database
+        # or constructed via UserProfile() if the request currently being
+        # handled is a User creation request (POST).
+        self.obj: UserProfile
+
+        super().__init__(obj, request)
+        self.subdomain = get_subdomain(request)
+        self.config = settings.SCIM_CONFIG[self.subdomain]
+
+        self._email_new_value: Optional[str] = None
+        self._is_active_new_value: Optional[bool] = None
+        self._full_name_new_value: Optional[str] = None
+        self._password_set_to: Optional[str] = None
+
+    def is_new_user(self) -> bool:
+        return not bool(self.obj.id)
+
+    @property
+    def display_name(self) -> str:
+        """
+        Return the displayName of the user per the SCIM spec.
+
+        Overriden because UserProfile uses the .full_name attribute,
+        while the superclass expects .first_name and .last_name.
+        """
+        return self.obj.full_name
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Return a ``dict`` conforming to the SCIM User Schema,
+        ready for conversion to a JSON object.
+        """
+        if self.config["name_formatted_included"]:
+            name = {
+                "formatted": self.obj.full_name,
+            }
+        else:
```

Comment:
I *really* don't like block, but Okta operates with a first_name, last_name model and doesn't provide a full name - which just doesn't work with our approach of having only a full_name. So I had to add some naive conversion.